### DISCUSSION TITLE
Updating Serenity from 6.5.2 to 6.6.0

### DISCUSSION THREAD

**Discussion by DGoertzen** [serenity-is/Serenity#6731] [Help] [answered]
I tried to update Serenity from 6.5.2 to 6.6.0 today, but I get a massive amount of typescript errors when I update Serenity.Net.Web (>300).   The errors are almost all of the form

`Error	TS2305	(TS) Module '"../"' has no exported member 'LanguageService'.	`

I didn't see anything in the release notes about what I should change in my code.  I suspect the errors are related to the change:

```
  - Renamed all service endpoint controllers so that they use Endpoint.cs suffix
  - ServiceEndpoint now derives from ControllerBase instead of Controller as services should not normally return views. If you returned view from a service endpoint for any reason, you'll need to move it to a controller. The endpoints generated by Sergen will have Endpoint suffix instead of Controller suffix. Similarly, pages will have Page suffix
```
Any suggestions on what I should change in my code to correct the errors?

Thanks!

**Comment by edwardch** [comment]
I had the same, looks like you can just rename like so: https://github.com/serenity-is/common-features/commit/a96e0535b9f353a3bcd14368cd35a86194a01c7d

I did a find & replace on `Controller : Controller` to `Page : Controller`, and `Controller : ServiceEndpoint` to `Endpoint : ServiceEndpoint`.
You'll need to update a couple of things manually, and your navigation files.

**Comment by volkanceylan** [comment, marked as answer]
If you updated Serenity packages, you should also dotnet tool update sergen and also don't forget to update Serenity.Pro.Coder as well and restart Visual Studio.

**Reply by Vinith-kumar47** [reply]
After succesfully updating the serenity.web and serenity.scripts and serenity.web.assets in the project ,while updating "dotnet tool update sergen" in the command prompt, it shows errors like below,
    Cannot find a manifest file.
    For a list of locations searched, specify the "-d" option before the tool name.
    If you intended to perform the action on a global tool, use the `--global` option for the command

How to rectify this error??

**Comment by volkanceylan** [comment]
Also there is no such requirement to change suffixes to Controller to Page/Endpoint. It is recommended so that the file name matches the controller class name, but not mandatory. The issue with what @DGoertzen had is either he only updated the Serenity.Net.Web and did not update Serenity.Pro.Coder and sergen, or did not restart Visual Studio. Old source generator looks for Controller base class, but the new ServiceEndpoint derives from ControllerBase, not Controller, so the old generator thinks it is not a service.

**Reply by DGoertzen** [reply]
I did not update sergen.  I will be able to test later today to see if that is the issue, but based on your description, it sounds like it is.  Thanks for the help.

**Reply by DGoertzen** [reply]
I updated Serenity.Pro.Coder and all the errors went away just as @volkanceylan indicated.  I'd initially missed upgrading it as I had my package source pointed to nuget.org, not serenity.is. I foolishly only upgraded the packages available on nuget.org. 

However, after I upgraded everything to 6.6.0, I was still getting one warning:

`Warning	CS8032	An instance of analyzer Serenity.Pro.Coder.EsmEntryPointsSourceGenerator cannot be created from XXX\6.6.0\analyzers\dotnet\cs\Serenity.Pro.Coder.dll: Could not load type 'Serenity.Pro.Coder.EsmEntryPointsSourceGenerator' from assembly 'Serenity.Pro.Coder, Version=6.5.2.0, Culture=neutral, PublicKeyToken=null'.`

I restarted visual studio (just as @volkanceylan said above) and the warning went away.  So all good now!  Thanks.

**Comment by LSNicholls** [comment]
If you branch and upgrade to 6.6.0 (or any other version -- this is a general question) and upgrade sergen, but want to switch between two branches with different versions, do you need to downgrade sergen to work on the original branch, or will the later version be okay for both?  Thanks in advance.

**Reply by VictorTomaili** [reply]
You can use both of the sergen versions by installing dotnet tool as project dependency instead of global to ensure do not break any functionality. But mostly, we would rather not break backward compatibility, and it should work with the latest sergen and pro coder version. 
https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools

**Comment by LSNicholls** [comment]
Thanks Victor. Thanks for confirming about backward compat; I really wasn't sure about this with a code generator tool.

I do understand about project versus global installation; however I am still not clear that this would actually work when I branched, considering it's in the same directory (obviously).  Nothing in the project file points to a sergen version.  Pro Coder, yes, but not sergen.

But I guess I just would need to change this:

`  <DotNetSergenDir>$(SerenitySrc)Serenity.Net.CodeGenerator\bin\</DotNetSergenDir>`

**Reply by VictorTomaili** [reply]
Check the .config/dotnet-tools.json file in your solution for sergen.

**Reply by LSNicholls** [reply]
Oh!!! Thank you so much.
